<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <title>Cufflinks - Transcript assembly, differential expression, and
          differential regulation for RNA-Seq</title>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
	<meta name="description" content="Open source tools for transcript assembly, differential expression, and
        differential regulation for RNA-Seq">
    <link rel="stylesheet" type="text/css" href="css/style.css"
      media="screen">
  </head>
  <body>
    <div id="wrap">
      <div id="top">
        <div class="lefts">
          <table cellpadding="2" width="100%">
            <tbody>
              <tr>
                <td> <a href="./index.html">
                    <h1>Cufflinks</h1>
                  </a>
                  <h2>Transcript assembly, differential expression, and
                    differential regulation for RNA-Seq</h2>
                </td>
                <td align="right" valign="middle"> 
                    <a href="http://bio.math.berkeley.edu/">
                    <img style="vertical-align:middle;padding-top:4px"
                      src="images/UCBerkeley-seal.scaled.gif" border="0">
                    </a>&nbsp;
                    <a href="http://genomics.jhu.edu/">
                    <img style="vertical-align:middle;padding-top:4px"
                      src="images/JHU-seal.gif" border="0">
                    </a>&nbsp;

                  <a href="http://www.cbcb.umd.edu/"><img
                      style="vertical-align:middle;padding-top:4px"
                      src="images/cbcb_logo.gif" border="0"></a>&nbsp;&nbsp;

                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div id="subheader">
        <table width="100%">
          <tbody>
            <tr>
              <td> <strong>Cufflinks</strong> assembles transcripts,
                estimates their abundances, and tests for <b>differential
                  expression</b> and <b>regulation</b> in RNA-Seq
                samples. It accepts aligned RNA-Seq reads and assembles
                the alignments into a parsimonious set of transcripts.
                Cufflinks then estimates the relative abundances of
                these transcripts based on how many reads support each
                one, taking into account biases in library preparation
                protocols. <br>
                <br>
                Cufflinks is a collaborative effort between the <a
                  href="http://bio.math.berkeley.edu/">Laboratory for
                  Mathematical and Computational Biology</a>, led by
                Lior Pachter at UC Berkeley, Steven Salzberg's 
                <a href="http://genomics.jhu.edu">computational genomics group</a>
                 at the Institute of Genetic Medicine at Johns Hopkins University, and
                <a href="http://woldlab.caltech.edu/">Barbara Wold's lab</a>
                at Caltech. <br>
                <br>
                Cufflinks is provided under the OSI-approved <a
                  href="http://en.wikipedia.org/wiki/Boost_Software_License">Boost
                  License</a> </td>
              <td align="right" valign="middle"> <a
                  href="http://opensource.org"><img alt="Open Source
                    Software" src="images/osi-certified.gif" border="0"></a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div id="main">
        <div id="rightside">
	
          <h2>Site Map</h2>
          <div class="box">
            <ul>
              <li><a href="index.html">Home</a></li>
              <li><a href="tutorial.html">Getting started</a></li>
              <li><a href="manual.html">Manual</a></li>
              <li><a href="howitworks.html">How Cufflinks works</a></li>
			  <li><a href="igenomes.html">Index and annotation downloads</a></li>
              <li><a href="faq.html">FAQ</a></li>
			  <li><a href="http://www.nature.com/nprot/journal/v7/n3/full/nprot.2012.016.html">Protocol</a></li>
			  <li><a href="report.html">Benchmarking</a></li>
            </ul>
          </div>

          <h2><u>News and updates</u></h2>
          <div class="box">
            <ul>
              <table width="100%">
                <tbody>
                  <tr>
                    <td>New releases and related tools will be announced
                      through the <a
                        href="https://lists.sourceforge.net/lists/listinfo/bowtie-bio-announce"><b>mailing
                          list</b></a></td>
                  </tr>
                </tbody>
              </table>
            </ul>
          </div>
          <h2><u>Getting Help</u></h2>
          <div class="box">
            <ul>
              <table width="100%">
                <tbody>
                  <tr>
                    <td>Questions about Cufflinks and Cuffdiff should be posted on our <a href="https://groups.google.com/forum/#!forum/tuxedo-tools-users"><b>Google Group</b></a>. Please use <a
                        href="mailto:tophat.cufflinks@gmail.com">tophat.cufflinks@gmail.com</a> for private communications only.
                      Please do not email technical questions to
                      Cufflinks contributors directly.</td>
                  </tr>
                </tbody>
              </table>
            </ul>
          </div>
       
          <a href="./downloads">
            <h2><u>Releases</u></h2>
          </a>
          <div class="box">
            <ul>
              <table width="100%">
                <tbody>
                  <tr>
                    <td>version 2.2.0</td>
                    <td align="right">5/25/2014</td>
                  </tr>
                  <tr>
                    <td><a href="./downloads/cufflinks-2.2.0.tar.gz"
                        onclick="javascript:
                        pageTracker._trackPageview('/downloads/cufflinks_source');
                        ">&nbsp;&nbsp;&nbsp;Source code</a></td>
                  </tr>
                  <tr>
                    <td><a
                        href="./downloads/cufflinks-2.2.0.Linux_x86_64.tar.gz"
                        onclick="javascript:
                        pageTracker._trackPageview('/downloads/cufflinks');
                        ">&nbsp;&nbsp;&nbsp;Linux x86_64 binary</a></td>
                  </tr>
                  <tr>
                    <td><a
                        href="./downloads/cufflinks-2.2.0.OSX_x86_64.tar.gz"
                        onclick="javascript:
                        pageTracker._trackPageview('/downloads/cufflinks');
                        ">&nbsp;&nbsp;&nbsp;Mac OS X x86_64 binary</a></td>
                  </tr>
                </tbody>
              </table>
            </ul>
          </div>
          


		  <h2>Related Tools</h2>
          <div class="box">
            <ul>
                <li><a href="http://monocle-bio.sourceforge.net">Monocle</a>:
                  Single-cell RNA-Seq analysis</li>
				<li><a href="http://compbio.mit.edu/cummeRbund/">CummeRbund</a>:
	                Visualization of RNA-Seq differential analysis</li>
              <li><a href="http://tophat.cbcb.umd.edu/">TopHat</a>:
                Alignment of short RNA-Seq reads</li>
              <li><a href="http://bowtie.cbcb.umd.edu">Bowtie</a>:
                Ultrafast short read alignment</li>
            </ul>
          </div>


          <h2>Publications</h2>
          <div class="box">
            <ul>
              <li style="font-size: x-small; line-height: 130%">
                <p>Trapnell C, Williams BA, Pertea G, Mortazavi AM, Kwan
                  G, van Baren MJ, Salzberg SL, Wold B, Pachter L.<b> <a
                      href="http://dx.doi.org/10.1038/nbt.1621">Transcript
                      assembly and quantification by RNA-Seq reveals
                      unannotated transcripts and isoform switching
                      during cell differentiation</a></b> <br>
                  <i><a href="http://www.nature.com/nbt">Nature
                      Biotechnology</a></i> doi:10.1038/nbt.1621</p>
                <br>
              </li>
              <li style="font-size: x-small; line-height: 130%">
                <p>Roberts A, Trapnell C, Donaghey J, Rinn JL, Pachter
                  L.<b> <a
                      href="http://genomebiology.com/2011/12/3/R22/abstract">Improving
                      RNA-Seq expression estimates by correcting for
                      fragment bias</a></b> <br>
                  <i><a href="http://www.genomebiology.com">Genome
                      Biology</a></i> doi:10.1186/gb-2011-12-3-r22</p>
                <br>
              </li>
              <li style="font-size: x-small; line-height: 130%">
                <p>Roberts A, Pimentel H, Trapnell C, Pachter
                  L.<b> <a
                      href="http://bioinformatics.oxfordjournals.org/content/early/2011/06/21/bioinformatics.btr355.abstract">
                      Identification of novel transcripts in annotated genomes using RNA-Seq</a></b> <br>
                  <i><a href="http://bioinformatics.oxfordjournals.org/">Bioinformatics</a></i> doi:10.1093/bioinformatics/btr355</p>
                <br>
              </li>
			  <li style="font-size: x-small; line-height: 130%">
                <p>Trapnell C, Hendrickson D,Sauvageau S, Goff L, Rinn JL, Pachter L<b> <a
                      href="http://dx.doi.org/10.1038/nbt.2450">Differential 
					 analysis of gene regulation at transcript resolution with RNA-seq
					</a></b> <br>
                  <i><a href="http://www.nature.com/nbt">Nature
                      Biotechnology</a></i> doi:10.1038/nbt.2450</p>
                <br>
              </li>
            </ul>
          </div>
          <h2>Contributors</h2>
          <div class="box">
            <ul>
              <li><a href="http://www.cs.umd.edu/%7Ecole/">Cole Trapnell</a></li>
              <li><a href="http://www.cs.berkeley.edu/%7Eadarob/">Adam
                  Roberts</a></li>
              <li>Geo Pertea</li>
			  <li>David Hendrickson<li>
			  <li>Loyal Goff</li>
			  <li>Martin Sauvageau</li>
              <li>Brian Williams</li>
              <li><a href="http://wormlab.caltech.edu/members/">Ali
                  Mortazavi</a></li>
              <li>Gordon Kwan</li>
              <li>Jeltje van Baren</li>
			  <li><a href="http://www.rinnlab.com">John Rinn</a></li>
              <li><a href="http://www.cbcb.umd.edu/%7Esalzberg/">Steven
                  Salzberg</a></li>
              <li><a href="http://biology.caltech.edu/Members/Wold">Barbara
                  Wold</a></li>
              <li><a href="http://www.math.berkeley.edu/%7Elpachter/">Lior
                  Pachter</a></li>
            </ul>
          </div>
          <h2>Links</h2>
          <div class="box">
            <ul>
              <li><a href="http://bio.math.berkeley.edu/">Berkeley LMCB</a></li>
              <li><a href="http://www.cbcb.umd.edu/">UMD CBCB</a></li>
              <li><a href="http://woldlab.caltech.edu/">Wold Lab</a></li>
            </ul>
          </div>
        </div>
        <!-- End of "rightside" -->
        <div id="leftside">
			<h2>2.2.0 release - 3/25/2014</h2>
			<p>This release introduces some new features designed to simplify and speed up Cufflinks workflows. Release version 2.2.0 includes two new programs, <a href="manual.html#cuffquant"><tt>cuffquant</tt></a> and <a href="manual.html#cuffnorm"><tt>cuffnorm</tt></a> that make it easier to quantify gene expression in experiments with many samples. These are particularly helpful for single cell RNA-Seq experiments, where the reads for each cell are provided as a separate FASTQ file or pair of files.
			<p>Release 2.2.0 also introduces <a href="manual.html#sample_sheets"><strong>sample sheets</strong></a> and <a href="manual.html#contrast_files"><strong>contrast files</strong></a>. These facilities make it easier to work with large analyses involving many samples.
			<p>Cuffquant quantifies gene and transcript expression levels for a single BAM file.  These levels are stored in a new binary file type, the CXB file.  You can then provide CXB files for your samples directly to Cuffdiff instead of BAMs.  Mixing BAM and CXB files is not yet supported.  Because expression levels for each sample are quantified by Cuffquant, Cuffdiff doesn't have to perform this step, which speeds up Cuffdiff runs substantially and lowers their memory footprints. We recommend that most users switch to this new Cuffdiff workflow for all experiments that involve more than a few samples.  However, note that running Cuffquant prior to running Cuffdiff is <strong>optional</strong> - you can still directly supply BAM files to Cuffdiff.
				</p>
			<p>Cuffquant files can also be passed to Cuffnorm, which simply computes a normalized table of expression values for genes and transcripts.  Unlike Cuffdiff, Cuffnorm performs no differential expression testing.  Cuffnorm also does not calculate confidence intervals on expression values.  Cuffnorm is meant to be used when all you want is a set of comparable expression values for genes, transcripts, CDS groups, and TSS groups, such as when you simply want to make heatmaps or dendrograms or plot values for individual genes.  Cuffnorm can output values in one of two formats: Cuffdiff tracking tables (which can be used with <a href="http://compbio.mit.edu/cummeRbund/">CummeRbund</a>), or in simple, tab-delimited tables.  The latter is the recommended output format for use with <a href="http://monocle-bio.sourceforge.net/">Monocle</a> in single cell RNA-Seq experiments.
			</p>
			<p> Here's an overview of the new (optional) workflow, compared to the old workflow described in our Nature Protocols paper:<img style="vertical-align:middle;padding-top:4px; display: block; margin-left: auto; margin-right: auto;"
          border=0 src="images/tuxedo_workflow.png">
				
			<br/>
			<br/>
			<h2>2.1.1 release - 4/11/2013</h2>
			<p>This release fixes a bug introduced with version 2.1.0, which for some users caused Cufflinks to output empty transcript GTF files.
				</p>
			<br/>
			<h2>2.1.0 release - 4/10/2013</h2>
				<p>
					This release substantially improves the accuracy, speed, and memory footprint of Cufflinks and Cuffdiff.  It is recommended for all users. Those who wish to see the impact of the accuracy improvements should look at the new <a href="report.html">benchmarking</a> section.  In addition to numerous bugfixes, the main changes are as follows:
				<ul>
					<li>Cuffdiff now includes a new statistical test.  Prior versions used a delta method-based test, which lacked power for designs with more than a few replicates.  The new test directly samples from the beta negative binomial model for each transcript in each condition in order to estimate the null distribution of its log fold change under the null hypothesis.  This test is substantially more powerful, resulting in improved accuracy over all experimental designs, particularly those with more than three replicates. A similarly improved test is now used for isoform-switching. The <a href="report.html">benchmarking</a> page shows the improvements in detail. 
						<li>Prior versions of Cuffdiff reported the FPKM for each gene and transcript that maximizes the joint likelihood of the reads from all replicates pooled together.  In version 2.1, Cuffdiff instead reports the mean of the maximum likelihood estimates from each replicate processed independently.  As shown in the benchmarking section, these two methods report nearly identical values.  However, the new method is faster and simpler to compute, and will enable new features for future releases.
					<li>The high and low confidence intervals reported by Cufflinks and Cuffdiff are now constructed from the samples generated from the beta negative binomial model, rather than estimated as twice the standard deviation.  This better reflects the underlying distribution of the FPKM.
					<li>The library normalization system in Cuffdiff 2 has been overhauled, and several new normalization-related options have been added:
						<ul>
							<li>The new <tt>--library-norm-method</tt> option now sets which method should be used to compute scaling factors for the library sizes.  The default method <tt>geometric</tt> is the same as prior releases of Cuffdiff (and the same as DESeq). The optional modes <tt>quartile</tt> and <tt>classic-fpkm</tt> are also supported.
							<li>The new <tt>--dispersion-method</tt> option controls how the variance model should be computed for each condition.  The default mode <tt>pooled</tt> computes a mean-variance model for each condition that has multiple replicates, averages these models to generate a "pooled" average, and uses it for all conditions.  This policy is borrowed from DESeq. Alternative models <tt>blind</tt> and <tt>per-condition</tt> are also supported. Prior versions of Cuffdiff used the method <tt>per-condition</tt>.
							<li>Several bugs for quartile normalization have been fixed.
							<li>Quartile normalization is no longer supported in Cufflinks, just in Cuffdiff.  Cufflinks only supports the <tt>classic-fpkm</tt> mode. 
							<li>All library size normalization is now conducted through the internal scaling factor. The external scaling factor should always be set to 1.0.
							<li>Library sizes and dispersions are now computed only on fragment counts from compatible fragments.  Prior versions counted intronic and other structurally incompatible fragments in some sections of the code.
						</ul>
					</li>
					<li>An optimized sampling procedure drastically improves running time for Cuffdiff.  Cufflinks also benefits from this change.  The improvements are particularly noticeable on deeply sequenced libraries.
					<li>The range of p-values that users should expect from Cuffdiff has changed.  Because the test is now based on explicit sampling from the beta negative binomial, users will not see values less than 10^-5 by default. The <tt>test_stat</tt> field of Cuffdiff's output still contains the  delta method's test statistic, but this test statistic is not used to compute p-values.  It is preserved for backward compatibility with some functions in <a href="http://compbio.mit.edu/cummeRbund/">CummeRbund</a>.
					<li>Some extraneous temporary output files have been removed after Cuffmerge runs.
				</ul>	
				</p>
				<br/>
				
			<h2>New google group for TopHat, Cufflinks, and CummeRbund users - 4/11/2013</h2>
			<p>To better handle the many questions we receive from users, we have launched a new google group for discussion and questions about our tools. We are simply unable to handle the email we receive, and our hope is that the users of of the group will help each other.  While we will continue to answer email sent to <a href="mailto:tophat.cufflinks@gmail.com">tophat.cufflinks@gmail.com</a> as we have time, we will also answer questions posted to the group, and the latter questions will be prioritized.  Please post questions or comments to the group (if your question hasn't already been asked and answered) before emailing <a href="mailto:tophat.cufflinks@gmail.com">tophat.cufflinks@gmail.com</a>.  
			</p>
			<br/>
			
			<h2>TransDecoder: calling novel ORFs in RNA-Seq assemblies- 4/11/2013</h2>
			<p>Our friend <a href="http://www.broadinstitute.org/~bhaas/">Brian Haas</a> (of <a href="http://trinityrnaseq.sourceforge.net/">Trinity</a> fame) has written a tool that might be helpful to Cufflinks users wishing to discover new coding genes and splice variants.  TransDecoder takes as input an assembly of RNA-Seq reads from Trinity, Cufflinks, and other tools and produces predictions about likely open reading frames as output.  http://transdecoder.sourceforge.net/
			</p>
			<br/>
			<h2>Cuffdiff 2 manuscript published at <em>Nature Biotechnology</em> - 12/11/2012</h2>
				<p>
				We are happy to announce a <a href="http://dx.doi.org/10.1038/nbt.2450">new manuscript</a> describing Cuffdiff 2 
				has appeared in print at <em>Nature Biotechnology</em>. The manuscript
				contains performance assessments of Cuffdiff 2 and a detailed 
				comparison of our methods against other tools for differential
				analysis.
				</p>
				<br/>
	
			<h2>2.0.2 release - 7/8/2012</h2>
				<p>This release fixes several bugs:
				<ul>
					<li>Some users were experiencing a crash on exit in Cufflinks when run with bias correction. The source of the crash has been fixed.
					<li>A few minor fixes in the estimation routines for cross-replicate variability.
					<li>Providing the same BAM file multiple times was producing inconsistent expression values.  This has been corrected.
				</ul>
				</p>
				<br/>
	
			<h2>2.0.1 release - 6/15/2012</h2>
				<p>This release addresses several bugs, most of which were introduced as 
				part of the 2.0.0 release:
				<ul>
					<li>Some users were experiencing a problem with the Linux 
						pre-compiled binary packages where Cuffdiff was reporting
						many genes and transcripts as having expression levels of
						zero or "nan" or "inf".  This was determined to be caused 
						by a recent change in the binary packaging process we use 
						to distribute pre-compiled binaries.  
					<li>Multi-read correction in Cuffdiff 2.0 was not always being
						applied when the user specified <tt>-u/--multi-read-correct</tt>.
					<li>Some rarely encountered GTF parsing issues have been corrected.
					<li>Effective length correction can be disabled in Cufflinks 
						and Cuffdiff with the new <tt>--no-effective-length-correction</tt> option.
				</ul>
				</p>
				<br/>
			
			<h2> Illumina iGenomes now include both Bowtie1 and Bowtie2 indexes - 5/23/2012</h2>
				  <br/>
				
			<h2>2.0.0 release - 5/4/2012</h2>
			  <p>This release substantially improves the accuracy and robustness of 
				differential analysis with Cuffdiff. The update also  
				resolves several user-reported issues and bugs, and several requested features.  
				Due to the large number of enchancements and fixes, users are 
				encouraged to treat this as a beta release.  A manuscript 
				describing the algorithmic improvements to the software is in preparation. Changes include:
				<ul>
					<li>Cuffdiff now reports estimated counts assigned to each gene and transcript, 
						along with count variances due to uncertainty and cross-replicate variability. 
						See the <a href="manual.html#count_tracking_format">manual</a> for 
						more details on the new count tracking file format. 
						Count tracking is not yet available in Cufflinks (e.g. with the -G option), 
						but this functionality will be ported over in a future release.
				    <li>Cuffdiff now tracks and records per-replicate FPKMs and counts. See the manual for
						more details on the new replicate tracking files. A version of 
						<a href="http://compbio.mit.edu/cummeRbund/">CummeRbund</a> 
						that exposes this information in many plot types will be forthcoming.
					<li>Cuffdiff reports replicate and run metadata as part of each run.
					<li>Some users were reporting a high FAIL rate on gene and
						transcripts quantification.  This has been resolved 
						according to a battery of tests using real and simulated data.  
						The root cause was that in conditions with substantial
						overdispersion across replicates, the FPKM variance-covariance 
						matrices  produced by the Cuffdiff model were not always 
						positive-definite.  Cuffdiff was detecting this, 
						and marking those genes as having unreliable confidence intervals. 
						Prior to 2.0.0, the model contained a heuristic approximation of the 
						covariances between assigned fragment counts (which are necessary 
						for calculating the variance on each gene's expression level),
						and this approximation was producing poorly conditioned matrices. 
						We have replaced the heuristic approximation with a direct sampling 
						approach, in effect "simulating" the assignment of fragments 
						to each isoform many times for each gene.  
						By simulating fragment generation and assignment to each transcript, 
						we are reconstructing variance-covariance matrices for assigned fragment 
						counts that are always properly condition.  This sampling 
						approach produces more accurate estimates of variance and covariance as well, 
						improving accuracy of transcript and gene level differential analysis. 
						Users should expect more accurate quantification and shorter, 
						more conservative lists of differentially expressed genes and trasncripts. 
					<li>After substantial performance testing, we have determined that
						the false discovery rate of Jensen-Shannon-based tests 
						(differential splicing, CDS switching, and promoter switching) can be 
						unnacceptably high when used with fewer than three replicates in the
						conditions being compared.  Cuffdiff now refrains from 
						performing significance tests when one of the conditions involved has 
						fewer than three replicates.  You can change this behavior with the new
						<tt>--min-reps-for-js-test</tt> option.  Cuffdiff still produces
						<tt>splicing.diff</tt>, <tt>cds.diff</tt>, and <tt>promoters.diff</tt> 
						regardless of how many replicates you have.  These files will include
						the JS distance scores, but none of the genes will be marked significant if you
						have fewer than the required number of replicates.
					<li>Cufflinks and Cuffdiff can now be told to ignore fragments that map 
						to the genome more than a specified number of times using the 
						<tt>--max-frag-multihits</tt> option.  By default, Cufflinks 
						and Cuffdiff still consider all fragments in the alignment file.
					<li>Cufflinks by default doesn't report assembled transfrags that 
						are built mostly from multiply mapping reads.  
						This behavior can now be controlled or disabled with the 
						new <tt>--max-multiread-fraction</tt> option.
					<li>Cufflinks by default fills small gaps in coverage when assembling transcripts.  
						Gaps smaller than 50bp are filled and the transfrags joined.  
						This behavior can be controlled or disabled with the new 
						<tt>--overlap-radius</tt> option.
					<li>Before testing for differential expression or regulation of genes and transcripts,
						Cuffdiff now checks that the variance model for the gene or transcript in question is a good fit.  
						This behavior can be controlled or disabled with the <tt>--min-outlier-p</tt> option.
						See the relevant section in the "How Cufflinks works" page for more on this.
					<li>A few bugs in the bias correction code and isoform deconvolution routines have been fixed, improving transcript-level expression accuracy.</li>
					<li>Positional bias correction was reducing accuracy on 
						certain datasets in some genes, so we have changed the
						 default bias correction algorithm to model sequence-specific 
						bias only.  Bias correction is still disabled by default, 
						and positional bias correction is still available as an 
						optional mode.</li>
					<li>Several minor issues related to library size normalization have been fixed.
					<li>A new library size normalization mode based on the 
						geometric mean has been added, and is now the default 
						for Cuffdiff. This method was introduced by Anders and
						 Huber (Genome Biology, 2010).
					<li>Cuffdiff now uses the <a href="http://eigen.tuxfamily.org">Eigen</a> 
						linear algebra libraries.  Eigen is a fast package for 
						matrix operations and makes good use of vector registers 
						in modern processors, speeding up some of the numerical routines used during abundance estimation.
					<li>Cuffdiff now requires Boost version 1.47 or later.
				</ul>
			  <br/>
	
			<h2>TopHat and Cufflinks protocol published at Nature Protocols - 3/12/2012</h2>
			  <p>A complete <a href="http://www.nature.com/nprot/journal/v7/n3/full/nprot.2012.016.html">bioinformatic protocol</a> for analysis of RNA-Seq data using our tools 
				has been published at <em>Nature Protocols</em>.  The protocol covers read alignment with TopHat, 
				gene and transcript discovery with Cufflinks, annotation analysis with Cuffmerge and Cuffcompare, 
				differential expression analysis with Cuffdiff, and visualization with CummeRbund.  
				Several variants of the protocol are included for those who wish to forgo certain analysis steps, such as gene discovery.
			  </p>
			  <br/>
			<h2>1.3.0 release - 1/2/2012</h2>
			  <p>This release improves the accuracy of Cuffdiff's isoform switching tests and fixes several bugs.
				<ul>
					<li>Cuffdiff find genes that are differentially spliced or switch promoters between conditions using the Jensen-Shannon distance metric. Previous versions of Cuffdiff tested for the signficance of observed shifts in relative isoform abundance using an analytic approximation of the variance of this metric for each gene.  However, when few replicates are available or sequencing is shallow, this approximation can be poor.  This release improves Cuffdiff's accuracy with a computational-derived estimate of the variance of the Jensen-Shannon metric by sampling in each gene.  This improvement substantially reduces the false positive rate of Cuffdiff's tests in <tt>splicing.diff</tt>, <tt>promoters.diff</tt>, and <tt>cds.diff</tt>.  Tests for changes in global expression are not affected.
					<li>A bug in Cuffmerge that caused a crash with a warning about improper sorting of SAM files has been corrected.
					<li>A bug that caused Cuffmerge to drop some reference transcripts from the output has been fixed.
					<li>A few minor issues with Cufflinks' pre-assembly alignment filters have been fixed.
				</ul>
			  <br/>
			<h2>1.2.0 release - 11/23/2011</h2>
			  <p>This release fixes a number of bugs and includes some signficant accuracy and performance improvements:
				  <ul>
				<li>Some users were experiencing a large number of genes or transcripts marked FAIL during Cuffdiff runs or Cufflinks quantification runs.  These were caused by one of several issues.  Most of these genes were due to numerical exceptions generated during importance sampling, a procedure originally intended to improve accuracy of abundance estimates in genes with one or more very low abundance isoforms.  After detailed simulation experiments, we have concluded that the gains in accuracy are minor and do not justify the number of genes that FAIL when this procudure can't be executed.  We have dropped both this routine and the bootstrap check, which can similarly generate FAIL genes under similar conditions. 
				<li>Improvements to FPKM variance estimates for genes and transcripts, resulting in better differential expression accuracy in Cuffdiff.
				<li>A bug that reduced accuracy in certain abundant genes has been corrected.
				<li>The <tt>--max-bundle-frags</tt> option, which skips extremely deeply sequenced loci that might otherwise crash the program, has been improved so that those loci are never allowed to fully load into memory.  This improves overall memory footprint and should prevent memory-related crashes.
				<li>A bug in bias correction that results in "Inf" and "NaN" values has been fixed.
				<li>An improved pre-quantification fragment collapse optimization has substantially lowered memory footprint and improved running time for Cuffdiff and Cufflinks abundance estimation.
				<li>A bug that halts runs on hash-collisions for read names has been fixed.
				<li>A buffer overflow in the SAM parser has been fixed.
				<li>Other minor bugfixes and accuracy improvements for quantification.
				</ul>
			  </ul>
			  </p>
			  <br/>
	
			<h2>CummeRbund released - 11/21/2011</h2>
			<p>
			Extracting biological insight from transcript-level RNA-Seq analysis can be very challenging. Due to the volume and complexity of output from RNA-Seq analysis pipelines, many users may choose to focus only on gene-level results, and thus miss crucial biological insights that a transcript-level analysis can provide. We are happy to present CummeRbund, an R/Bioconductor package that simplifies the organization, access, exploration, and visualization of the various output files of a Cuffdiff differential expression analysis. CummeRbund begins by re-organizing the Cuffdiff output files, and storing these data in a local SQLite database. During this process, CummeRbund indexes the data to speed up access to specific feature data (genes, isoforms, TSS, CDS, etc.), and preserves the various relationships between these features. Access to data elements is done through R via the RSQLite package and data are presented in appropriately structured R classes with various convenience functions designed to streamline your workflow. 
			<p>
				CummeRbund simplifies the way in which you access and analyze your RNA-Seq data. Features include:
				<ul>
				<li>Numerous plotting methods to allow for rapid visualization of RNA-Seq data quality and global statistics such as FPKM distribution, as well as simple routines for plotting expression levels for one or more genes, their isoforms, TSS groups, or CDS groups. 
				<ul>
			      <li>Plots are generally publication-ready.  However, because they are built with <a href="http://had.co.nz/ggplot2/">ggplot2</a>, plot objects returned by the plotting routines are easy to manipulate in a live R session so you can tweak them to your specification.
				</ul>
			  <li>Persistent storage and indexing of cuffdiff data in a relational database.
				<ul>
			     <li>Quicker searching for feature-specific information
			     <li>Seamless aggregation of all related data points.
			     <li>Direct SQL/RSQLite querying of cuffdiff data for access to complex datasets using specialized queries.
				</ul>
				<li>Formalized R classes for data access and manipulation.
					<ul>
			      <li>'Pointer' classes call data directly from SQL tables.
			      <li>'Data' classes contain queried results and can be manipulated directly in R.
			      <li>Methods for direct access to FPKM values, differential expression data, and additional annotation of all gene-, isoform-, TSS-, and CDS-level features.
			      <li>Output formats allow for standard browsing/analysis in R (data.frame, list, etc).
				</ul>
			  <li>Geneset-level data access:
				<ul>
			      <li>Relevant data for meaningful subsets of genes (e.g. 'significantly regulated') can be quickly aggregated and stored in a geneset object.
			      <li>plot wrappers for heatmaps, expression profiles, barplots, volcano plots, scatterplots, etc.
			 	</ul>
			<li>Gene-centric data access:
				<ul>
			     <li>All interconnected data for a single gene are aggregated into a single data object.
			     <li>Access to all relevant FPKM, differential expression, and annotation data for all features of a single gene.
			     <li>Gene-specific plots (expression profiles)
				</ul>
			 <li>Individual Feature-level data access as well:
				<ul>
			      <li>Individual Isoforms, TSS, CDS, etc. are also available.
			      <li>Direct access to distribution test results for alternative splicing, TSS groups, and CDS usage.
				</ul>
			</ul>

			CummeRbund is being made freely available under the OSI approved Artistic License 2.0 and can be downloaded from <a href="http://compbio.mit.edu/cummeRbund/">http://compbio.mit.edu/cummeRbund/</a>. CummeRbund has also been included as part of the new R/<a href="http://www.bioconductor.org">Bioconductor</a> v2.9 release and can be installed in a similar manner to standard Bioconductor packages.
	</p>
	<br/>
			<h2>1.1.0 release - 9/8/2011</h2>
			  <p>This is a fix release to address several issues reported by our users:
				  <ul>
				<li>Cuffdiff now includes a more sophisticated check for sufficient sequencing depth prior to testing for differences, which substantially improves the accuracy of differential expression analysis in loci with low to medium depth. 
				<li>A bug in the calculation of the parameters for the beta negative binomial distribution used during variance estimation has been fixed.
				<li>Some users have reported that Cufflinks or Cuffdiff crashes due to running out of memory when processing highly expressed loci. Both programs now have a maximum number of fragments that can fall within a locus.  If a locus has more than this maximum, it is skipped.  The threshold is configurable via the <tt>--max-bundle-frags</tt> option.
				<li>Gffread now keeps the longest CDS when collapsing transcripts.
				</ul>
				<p><b>FILE FORMAT CHANGES:</b>
				<ul>
				<li>By popular request, fold changes are now reported log2, rather than natural log
				<li>The status column in fpkm tracking files has been replaced by a per-condition quantification status column
				<li>The header size limit in Cufflinks' BAM parser used to have a 4 megabyte limit.  This has been removed to allow Cufflinks to be used on assemblies with many contigs.
			  </ul>
			  </p>
			  <br/>
			
			<h2>iGenomes index and annotation packages available for download - 7/31/2011</h2>
			  <p> Illumina has generously provided a set of freely downloadable packages that contain everything you need to get started working with TopHat and Cufflinks.  These packages contain Bowtie indexes for the human, mouse, and fly genomes as well as many others.  The packages also contain annotation files (in GTF format) from UCSC, Ensembl, NCBI, and other sources.  These files are augmented with the special attributes Cufflinks needs to perform differential splicing and promoter analysis.  We strongly encourage users to download and try these packages!
			  <p></p>
			  <br>

			<h2>Cufflinks RABT assembly paper published - 6/26/2011</h2>
			  <p> Our new <a href="http://bioinformatics.oxfordjournals.org/content/early/2011/06/21/bioinformatics.btr355.abstract">paper</a>
				on the Cufflinks reference annotation based transcript (RABT) assembly method introduced in v1.0.0
				has appeared in <i><a href="http://bioinformatics.oxfordjournals.org/">Bioinformatics</a></i>. 
				The paper describes how the RABT assembler builds upon a known reference annotation to better identify novel transcripts.
				You can try the RABT assembler in Cufflinks by using the <tt>-g</tt> option as explained in the manual.</p>
			  <p>Please cite this paper as well as the original Cufflinks paper if using the RABT assembler in your work.</p>
			  <p></p>
			  <br>
        
			<h2>1.0.3 release - 6/1/2011</h2>
			  <p>This is a fix release to address several issues reported by our users:
				  <ul>
				<li>Several Cuffmerge bugs have been fixed.  One of these caused Cuffmerge to fail with a message about chromosome sort ordering.
				<li>A comment parsing issue in Cuffmerge has been fixed.
				<li>GTF files emitted by Cuffcompare are now lexicographically sorted by chromosome.
				<li>Gffread now has some transcript clustering and redundancy reduction features, similar to those found in Cuffcompare
				<li>The header size limit in Cufflinks' BAM parser used to have a 4 megabyte limit.  This has been removed to allow Cufflinks to be used on assemblies with many contigs.
			  </ul>
			  </p>
			  <br/>
		  <h2>1.0.2 release - 5/22/2011</h2>
		  <p>This release fixes several bugs and adds a few enhancements:
	  	  <ul>
			<li>The differential splicing and promoter and cds use testing in 1.0.0 and later was broken, resulting in very few (often zero) differentially regulated genes. This has been fixed.
			<li>Many users were confused about the magnitude of quartile-normalized FPKM values in Cuffdiff.  Now, quartile normalized FPKM values are scaled by a constant such that they are numerically in roughly the same range as non-quartile normalized FPKMs.  Thanks to Ariel Schwartz for suggesting this change.
			<li>A performance issue when processing BAMs with many unmapped reads has been fixed.
			<li>Cuffcompare now creates <tt>.tmap</tt> and <tt>.refmap</tt> in the directory of the input file(s)
				with the same prefix as given by -o
			<li>An issue that caused cuffmerge to fail with a message about improperly sorted SAM files has been corrected.
			<li>Improved cuffcompare GFF support.
			<li>Cuffdiff and Cufflinks now accept new options controlling whether all hits are counted towards the FPKM denominator, or only those compatible with some transcript in the reference annotation.  Counting only compatible hits avoids certain types of bias that arise when one sample contains far more hits that aren't compatible with any transcript than the other sample does.  For example, if one sample contains vastly more mapped ribosomal RNA hits, FPKM values will appear lower in that sample, potentially leading to false positive differential expression calls.  Cuffdiff by default now uses only compatible hits.  Cufflinks still uses total hits by default, as using compatible hit accounting requires a reference GTF. 
		  </ul>
		  </p>
		  <br/>
		  <h2>1.0.1 release - 5/6/2011</h2>
		  <p>This fix release corrects several issues introduced with 1.0.0:
	  	  <ul>
			<li>The binary packages for 1.0.0 failed to include gtf_to_sam and gffread, causing cuffmerge to fail.
			<li>Cuffdiff's <tt>splicing.diff</tt> was missing the <tt>q_value</tt> column header
			<li>Several portability issues have been fixed (thanks to Nathan Weeks for the patches)
			<li>Cuffmerge's help message incorrectly listed several options.
		  </ul>
		  </p>
		  <br/>
          <h2>1.0.0 release - 5/5/2011</h2>
          <p>This release represents a huge leap for Cufflinks in terms
            of performance and features. It is highly recommended that
            all users upgrade to this version of Cufflinks. Updates and
            improvements include: </p>
          <ul>
            <li>A new Reference Annotation Based Transcript (RABT)
              assembly mode has been added. More details can be found in
              the <a href="howitworks.html#hmul">How Cufflinks Works</a>
              section. </li>
            <li>Major improvements to Cuffdiff. Handling of replicates in Cuffdiff have been
              dramatically overhauled. Cuffdiff now models fragment
              count overdispersion with a beta negative binomial distribution in each condition prior to testing.
              See the substantially updated page on <a
                href="howitworks.html#hdif">How Cufflinks works</a> for
              more details.</li>
            <li>Bias correction described <a
                href="http://genomebiology.com/2011/12/3/R22/">here</a>
              is now enabled with the <tt>-b/--frag-bias-correct</tt>
              option (<tt>-r/--reference-seq</tt> is no longer in use).
              A path to the reference multi-fasta used in mapping must
              be supplied following the option. </li>
            <li>Added support for improved handling multi-mapping reads.
              Enable with the <tt>-u/--multi-read-correct</tt> option.
              See <a href="howitworks.html#hmul">How Cufflinks Works</a>
              for more details. </li>
            <li>Trimming has been instituted to more accurately locate
              the 3' ends of transcripts during assembly based on
              coverage.</li>
            <li>Cufflinks now includes a new tool called Cuffmerge to help merge
              assemblies from multiple samples into a single GTF for use
              with Cuffdiff. The tool also helps integrate a reference
              annotation file. See the <a href="tutorial.html">Getting
                Started</a> page for more details. </li>
            <li>Output file formats have been made consistent between
              Cufflinks and Cuffdiff. See the <a href="manual.html">Manual</a>
              for more details on the new formats.</li>
            <li>Both GFF3 and GTF2.2 annotations are now fully supported
              as input to all programs (see <a href="gff.html">here</a>).</li>
            <li>Improved reporting of map properties.</li>
            <li>The programs now check for available updates automatically on launch.</li>
            <li>Upper-quartile normalizaion has been fixed to be consistent with published literature (enable with <tt>-N/--upper-quartile-norm</tt>).</li>
            <li>Fixed a bug where some splice-junction reads were lost
              in quantitation.</li>
            <li>Fixed a bug where reads landing in introns were
              over-filtered in assembly.</li>
            <li>Numerous improvements in speed for both assembly and
              quantitation.</li>
			<li>Cuffdiff now uses dramatically less memory.  Cufflinks' memory footprint has also shrunk.</li>
            <li>Numerous minor bug fixes.</li>
          </ul>
          <p></p>
          <p><br>
          </p>
          <h2>Cufflinks Bias Correction paper published - 3/16/2011</h2>
          <p> Our new <a href="http://genomebiology.com/2011/12/3/R22/">paper</a>
            on the Cufflinks bias correction method introduced in v0.9.0
            has appeared in <i><a href="http://www.genomebiology.com/">Genome
                Biology</a></i>. The paper describes the details of the
            method and provides validations for the improvements in
            expression estimates it produces. As a reminder, bias
            correction is only activated when a reference sequence
            (fasta) is supplied with the -r option. </p>
          <p>Please cite this paper if using Cufflinks bias correction
            in your work.</p>
          <p></p>
          <br>
          <h2>New "Getting Help" email address for TopHat and Cufflinks</h2>
          <p>In order to more effectively answer user help requests and
            improve usability and documentation, we have created an
            email address to which users can send messages for technical
            support. If you have questions about Cufflinks or TopHat,
            please send them to <a
              href="mailto:tophat.cufflinks@gmail.com">tophat.cufflinks@gmail.com</a>.
            We will do our best to answer your question in a timely
            fashion, although <b>please read the manual carefully</b>
            before sending your email. We have very limited time to
            answer questions, and most questions require careful,
            technical answers. </p>
          <p>If you believe you have found a bug in the software, please
            <b>include a small package of test data</b> with your email
            so that we can reproduce your problem locally. A test
            example makes it much easier to correct the issue. </p>
          <br>
          <h2>TopHat and Cufflinks now supported through Galaxy</h2>
          <p>We are very pleased to announce that you can now run TopHat
            and Cufflinks through <a href="http://usegalaxy.org">Galaxy</a>.
            The Galaxy project aims to make informatics tools accessible
            through the web, and allows you to experiment with parameter
            settings and create sophisticated analysis workflows easily.
            Galaxy is developed by researchers at Emory University and
            Penn State in the <a href="http://bx.mathcs.emory.edu/">Taylor</a>
            and <a href="http://www.bx.psu.edu/%7Eanton/">Nekrutenko</a>
            labs, respectively. We are extremely grateful for the Galaxy
            team's work, and proud to have TopHat and Cufflinks offered
            through their platform. </p>
          <br>
          <h2>0.9.3 release - 11/30/2010</h2>
          <p>This release fixes several issues that affect abundance
            estimation and differential expression accuracy, and is
            strongly recommended for all users. Additionally, there are
            some speed and threading improvements during bias modeling
            and correction. Finally, a bug which causes Cufflinks to
            crash on some BAM files has been fixed. </p>
          <p> <br>
          </p>
          <h2>0.9.2 release - 10/26/2010</h2>
          <p>This release modifies the way library types are handled to
            add support for more strand-specific protocols as well as
            adaptability to future protocols that may be introduced.
            Note that the way library-types are specified in the input
            has changed as a result. See the <a
              href="manual.html#library">Manual</a> for more details.</p>
          <p>Some portability issues, which resulted in segfaults on
            some systems, have been fixed in the precompiled binaries.</p>
          <br>
          <h2>0.9.1 release - 10/3/2010</h2>
          <p>This release includes two bug fixes and some enhancements
            for strand-specific RNA-Seq libraries: </p>
          <ul>
            <li>SAM files without headers, such as those produced by
              TopHat 1.0.14 and earlier, were being erroneously reported
              as unsorted during GTF-based quantification, causing
              Cufflinks and Cuffdiff to exit. This has been fixed, so
              that SAM files are now correctly handled, restoring
              compatibility with output from TopHat 1.0.14 and earlier.</li>
            <li>Cufflinks and Cuffdiff were sometimes producing
              different FPKM values for minor isoforms when run on the
              same sample. They are now consistent, and should agree on
              isoform abundance calls. </li>
            <li>Strand-specific, single end libraries are now supported.
              Some library type names have changed to reflect the
              expanded list of supported protocols.</li>
          </ul>
          <p></p>
          <br>
          <h2>Update - fix for 0.9.0</h2>
          <p>The binaries and source have been updated to address a
            floating point exception when using a SAM file with no
            header and without a GTF</p>
          <br>
          <h2>New developer - Adam Roberts</h2>
          <p>The Cufflinks team has been joined by Adam Roberts, a Ph.D.
            student from the UC Berkeley Department of Computer Science.
            Adam has made many improvements to Cufflinks (see the
            announcement below), and we are very fortunate to benefit
            from his talents and expertise.</p>
          <br>
          <h2>0.9.0 release - 9/27/2010</h2>
          <p>This release includes significant bug fixes as well as some
            major new features. Enhancements include: </p>
          <ul>
            <li>Thanks to the work of Adam Roberts, Cufflinks now
              automatically learns certain properties of each RNA-Seq
              library you analyze with it, including sequence specific
              biases. Transcript and gene abundance estimates are
              significantly more accurate when run with our (optional)
              new bias correction technology. A manuscript describing
              these changes is in preparation. </li>
            <li>Due to Adam's changes, we do <b>not</b> recommend that
              you supply the mean and standard deviations of the
              fragment length distribution for your library. Cufflinks
              now learns these automatically. </li>
            <li>Cuffdiff now allows you to supply reads from multiple
              technical or biological <b>replicates</b>, and adjusts
              transcript and gene abundance estimates and their
              confidence intervals accordingly.</li>
            <li>Cuffdiff will by default report testing results between
              all pairs of samples, rather than treating samples as a
              time series. Time series analysis is still supported via
              the <tt>--time-series</tt> option.</li>
            <li>Cufflinks and Cuffdiff now offer improved compatibility
              with reads aligned with Life Technology's BioScope
              aligner. Strand specific SOLiD RNA-Seq runs are fully
              supported.</li>
            <li>Cufflinks and Cuffdiff support quantile normalization of
              FPKM.</li>
            <li>Cufflinks and Cuffdiff now support both BAM and SAM
              files. Both file types require that the chromosome names
              in header <tt>@SQ</tt> records, if present, <b>must</b>
              be sorted in the same order as the reads.</li>
            <li>Cufflinks and Cuffdiff are much less verbose in their
              output by default, but each supports a "verbose" mode with
              a text progress bar for users that want to keep an eye on
              their runs</li>
            <li>In response to user requests, FPKM values are now
              provided by Cufflinks and Cuffdiff, even if their
              expression levels are zero, in an effort to make
              downstream analysis of Cuffdiff easier.</li>
            <li>The new "mask" GTF option allows you to supply Cufflinks
              with regions of the genome that it should ignore during
              assembly and quantification. Cuffdiff also supports mask
              GTF files for convenience.</li>
            <li>Significant assembler fixes and improvements.</li>
            <li>Some of the output files produced by Cuffdiff have
              updated and simplified formats.</li>
            <li>There are a number of new command line options that
              users may wish to set under certain circumstances.</li>
          </ul>
          <p>Note that building Cufflinks from source now requires the <a
              href="http://samtools.sourceforge.net">SAM tools</a>. See
            the installation instructions for more details. </p>
          <br>
          <h2>0.8.3 build update - 7/2/2010</h2>
          <p> The 0.8.3 release on 6/30 included a broken version of
            Cuffdiff. This and a few other minor issues have been
            resolved and the builds and source have been updated. </p>
          <br>
          <h2>0.8.3 release - 6/30/2010</h2>
          <p> Fixes and accuracy improvements to the Cufflinks
            assembler. This update is strongly recommended for users
            interested in identifying novel transcripts. Other changes:
          </p>
          <ul>
            <li>Additional expression reporting fixes</li>
            <li>Output from Cufflinks and Cuffdiff can now be directed
              to a separate directory.</li>
          </ul>
          <p></p>
          <br>
          <h2>Cufflinks paper published - 5/2/2010</h2>
          <p> Our <a href="http://dx.doi.org/10.1038/nbt.1621">paper</a>
            on Cufflinks has appeared at <i><a
                href="http://www.nature.com/nbt">Nature Biotechnology</a></i>.
            In a study of developing mouse muscle cells, we identified
            thousands of new transcripts and detected differential
            splicing and promoter use in hundreds of genes by examining
            changes in isoform abundance levels. The paper is aimed at a
            general audience, so much of the algorithmic and
            mathematical detail can be found in the <a
href="http://www.nature.com/nbt/journal/vaop/ncurrent/extref/nbt.1621-S1.pdf">supplemental
              methods</a> (WARNING: 2 MB PDF). </p>
          <br>
          <h2>0.8.2 release - 3/26/2010</h2>
          <p> Numerous bug fixes and significant performance
            improvements in Cufflinks and Cuffdiff. Note that some the <strong>formats</strong>
            of some files <strong>have changed</strong>. Other changes:
          </p>
          <ul>
            <li>Expression values were not being reported for a small
              number of genes when using certain annotation files.</li>
            <li>Fixed a divide-by-zero error in Cuffdiff</li>
            <li>Various GTF parser fixes.</li>
            <li><tt>.expr</tt> files produced by Cufflinks now contain
              FPKM confidence interval columns</li>
            <li>Isoform filtering was being innappropriately applied
              when working with annotation in Cufflinks (via <tt>--GTF</tt>)</li>
            <li>The <tt>.tmap</tt> files reported by Cuffdiff contained
              zero in FPKM field.</li>
            <li>New command line options <tt>--num-importance-samples</tt>
              and <tt>--max-mle-iterations</tt> allow advanced users to
              influence of the transcript abundance estimation. </li>
          </ul>
          <p></p>
          <br>
          <h2>0.8.1 release - 2/13/2010</h2>
          <p> This is a minor fix release. Cufflinks 0.8.0 had some
            lingering references to "RPKM", which have been replaced
            with "FPKM". Reported expression values have always been in
            FPKM, this is simply a nomenclature change to better reflect
            what RNA-Seq actually measures. See the section on FPKM in
            the description of how Cufflinks <a href="howitworks.html">works</a>
          </p>
          <br>
          <h2>0.8.0 release - 2/5/2010</h2>
          <p> We are happy to announce a major update to Cufflinks that
            introduces some powerful new features and includes a number
            of performance improvement and bug fixes. Highlights
            include: </p>
          <p> </p>
          <ul>
            <li>Cufflinks now includes a new tool, "Cuffdiff", which
              performs testing for differential expression, splicing,
              promoter use, and coding sequence output on two or more
              RNA-Seq samples. See the greatly expanded <a
                href="manual.html">manual</a> for details.</li>
            <li>Cuffcompare now reports a file containing the "union" of
              all transfrags in the files you give it as input, greatly
              simplifying downstream validatation of novel transcripts.</li>
            <li>Cufflinks' assembler has been overhauled and optimized,
              resulting in a speedup of 4-5 times over version 0.7.0,
              and a <b>greatly</b> reduced memory footprint. Phasing of
              splicing events has also been improved.</li>
            <li>Many bugfixes.</li>
          </ul>
          <br>
          <h2>0.7.0 release - 9/26/2009</h2>
          <p> The first public release of Cufflinks is now available for
            download. Cufflinks is a program for the comparative
            assembly of transcripts and the estimation of their
            abundances in an RNA-Seq experiment. It runs on Linux and OS
            X. It also comes with a tool to track transcripts across
            multiple samples, for example in a time course of RNA-Seq. </p>
          <p>Cufflinks takes as input a file of alignments in <a
              href="http://samtools.sourceforge.net">SAM</a> format, and
            reports transfrags in <a
              href="http://mblab.wustl.edu/GTF22.html">GTF</a> format.
            You can use <a href="http://tophat.cbcb.umd.edu">TopHat</a>
            to align your reads, as TopHat reports its alignments in SAM
            format. </p>
          <p>This software is a work in progress - it is a beta release,
            and new features will continue to be added over the next
            couple of weeks. To suggest a feature or report a bug,
            please email <a href="mailto:cole@cs.umd.edu">Cole Trapnell</a>
          </p>
        </div>
      </div>
      <div id="footer">
        <table cellspacing="15" width="100%">
          <tbody>
            <tr>
              <td> This research was supported in part by NIH grants
                R01-LM06845 and R01-GM083873, NSF grant CCF-0347992 and
                the Miller Institute for Basic Research in Science at UC
                Berkeley. </td>
              <td align="right"> Administrator: <a
                  href="mailto:cole@cs.umd.edu">Cole Trapnell</a>.
                Design by <a href="http://www.free-css-templates.com"
                  title="Design by David Herreman">David Herreman</a> </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <!-- Google analytics code -->
    <script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
    <script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6101038-2");
pageTracker._trackPageview();
} catch(err) {}</script>
    <!-- End google analytics code -->
  </body>
</html>
